src/dbnode/encoding/proto/custom_unmarshal.go

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+)
+
+var (
+	// Groups in the Protobuf wire format are deprecated, so simplify the code significantly by
+	// not supporting them.
+	errGroupsAreNotSupported = errors.New("use of groups in proto wire format is not supported")
+	zeroValue                unmarshalValue
+)
+
+type customFieldUnmarshaler interface {
+	sortedCustomFieldValues() sortedCustomFieldValues
+	nonCustomFieldValues() *dynamic.Message
+	numNonCustomValues() int
+	resetAndUnmarshal(schema *desc.MessageDescriptor, buf []byte) error
+}
+
+type customUnmarshaler struct {
+	schema *desc.MessageDescriptor
+
+	decodeBuf *buffer
+
+	customValues sortedCustomFieldValues
+
+	nonCustomValues *dynamic.Message
+	numNonCustom    int
+}
+
+func newCustomFieldUnmarshaler() customFieldUnmarshaler {
+	return &customUnmarshaler{
+		decodeBuf: newCodedBuffer(nil),
+	}
+}
+
+func (u *customUnmarshaler) sortedCustomFieldValues() sortedCustomFieldValues {
```

Comment:
could we name this `customFieldValues()` given that they're already sorted?

Reply:
@mway I prefer this just because its really important that this return sorted results so I want the API to make that clear

Reply:
sure, sgtm